### PR TITLE
Execute tools for streaming models on `Infrastructure.getDefaultWorker`

### LINKED ISSUE #845: Tool executions on streaming models lead to `The current thread cannot be blocked: vert.x-eventloop-thread-0`
When combining a streaming model with a tool, that does e.g. a rest call, it leads to the following error message: 

```
The current thread cannot be blocked: vert.x-eventloop-thread-0
```

Here is an example: 

```java

    @Singleton
    static class MyTool {

        @Tool(value = "the answer to life, the universe and everything")
        public String answer() {
            return Uni.createFrom().item("42")
                    .onItem().delayIt().by(Duration.ofMillis(500))
                    .await().indefinitely();
        }
    }

    @RegisterAiService(tools = MyTool.class)
    interface MyService {
        Multi<String> chat(String msg);
    }

```

The reason is, that 

https://github.com/quarkiverse/quarkus-langchain4j/blob/0cd6f69fc7092b67534e7d60ef2ec87df4d8f88d/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java#L572

tries to execute the tool on the event loop. 

I think this should be easily fixable by running it on `Infrastructure.getDefaultWorkerPool()`. 

I can contribute a fix, but implementing a test will not be possible without merging first #844. 

This is not to be confused withe the request for supporting `Uni` as in #837.

### PR DESCRIPTION
**Pull request by dastrobu**
This fixes `The current thread cannot be blocked: vert.x-eventloop-thread-0` on tool executions.

Closes #845

### CODE REVIEW

**Comment by geoand** [comment]
Thanks for looking into this.

I need to take a closer look as I am pretty sure this will cause trouble in some circumstances.

**Comment by dastrobu** [comment]
Sure, there is a unit test for reproducing the issue from #845 in [StreamingChatLanguageModelToolTest](https://github.com/dastrobu/quarkus-langchain4j/blob/execute-tool-on-infra/core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java)

This should help adressing the issue. Maybe you could provide test cases for those "circumstances" you have in mind, as the build is passing for this PR. Then I can look into the issues. 

I can cleanup this draft PR as soon as #844 is merged. Then it will also be easier to add test cases and fix the implementation.

**Comment by geoand** [comment]
I just merged https://github.com/quarkiverse/quarkus-langchain4j/pull/844 so indeed this now needs a rebase. 

I plan on reviewing this today, so I'll let you know soon enough

File: core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        private final MyService service;
+
+        MyResource(MyService service) {
+            this.service = service;
+        }
+
+        @GET
+        public Multi<String> streaming() {
+            return service.chat("what is the Answer to the Ultimate Question of Life, the Universe, and Everything?");
```
Comment (geoand) [inline core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java:48]: I would be interested in seeing which thread the items are emitted on when tools are used.

Comment (dastrobu) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java:48]: Currently, all calls are executed on the event-loop thread. 

So, waiting in a call to generate a response will also fail: 

```
    public static class MyModelSupplier implements Supplier<StreamingChatLanguageModel> {
        @Override
        public StreamingChatLanguageModel get() {
            return (messages, handler) -> {
                handler.onNext("4");

                // block thread intentionally to verify there is no: The current thread cannot be blocked: vert.x-eventloop-thread-1
                Uni.createFrom().voidItem().onItem().delayIt().by(Duration.ofMillis(500)).await().indefinitely();

                handler.onNext("2");
                handler.onComplete(new Response<>(new AiMessage("")));
            };
        }
    }
```

However, this is probably not something we need to support, as the API is callback-based.

A nicer implementation would be 

```
            return (messages, handler) -> {
                Multi.createFrom().items("4", "2").onCompletion().invoke(() -> {
                            handler.onComplete(new Response<>(new AiMessage("")));
                        })
                        .subscribe().with(handler::onNext);
```

This is different for tool calls, though, where a response needs to be returned blocking, as all calls after `handler.onComplete(Response.from(AiMessage.from(toolExecutionRequest)));` run synchronously on one thread. 

The question is, what is the benefit of running on the event loop?

Here are some logs, executing the test case from the main branch:

```
2024-09-02 15:19:20,973 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyModelSupplier$1] (vert.x-eventloop-thread-1) requesting tool execution
2024-09-02 15:19:20,976 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyTool] (vert.x-eventloop-thread-1) executing tool call
2024-09-02 15:19:20,983 ERROR [io.qua.lan.run.too.QuarkusToolExecutor] (vert.x-eventloop-thread-1) Error while executing tool 'class io.quarkiverse.langchain4j.test.StreamingChatLanguageModelToolTest$MyTool': java.lang.IllegalStateException: The current thread cannot be blocked: vert.x-eventloop-thread-1
```

And here, after the change to `Infrastructure.getDefaultWorkerPool()`:

```
2024-09-02 15:21:36,177 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyModelSupplier$1] (executor-thread-1) requesting tool execution
2024-09-02 15:21:36,180 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyTool] (executor-thread-1) executing tool call
2024-09-02 15:21:37,124 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyModelSupplier$1] (executor-thread-1) processing tool execution result
```

Note that there is already some offloading to `Infrastructure.getDefaultWorkerPool()` in [this line of code](https://github.com/quarkiverse/quarkus-langchain4j/blob/2b014289dd3f8c2d14bc6494bae76a0b66a0cbea/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java#L151).

After looking again through the code, I see the options of  
 * running `AiServiceTokenStream` on `Infrastructure.getDefaultWorkerPool` as proposed here. 
 * requesting an API extension from langchain to support an asynchronous API for tool executions.

Comment (dastrobu) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java:49]: I missed your comment https://github.com/quarkiverse/quarkus-langchain4j/pull/846#issuecomment-2324667832, let me check....

Comment (geoand) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java:49]: > Note that there is already some offloading to Infrastructure.getDefaultWorkerPool() in [this line of code](https://github.com/quarkiverse/quarkus-langchain4j/blob/2b014289dd3f8c2d14bc6494bae76a0b66a0cbea/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java#L151).

Indeed, but that happens before the `Multi` is created. 

> requesting an API extension from langchain to support an asynchronous API for tool executions.

Yeah, this is a far better option as `running AiServiceTokenStream on Infrastructure.getDefaultWorkerPool as proposed here` is not the proper way to go about this

Comment (dastrobu) [reply core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java:49]: Resolving this conversation in favor of the main conversation on the PR.

**Comment by geoand** [comment]
I'm glad it works!

I agree it's not a great solution and I want to put a little more thought into what can be done for this situation

**Pull request closed without merging**